kickstarter/android-oss
# EP-424: Page Viewed (search)

# 📲 What

Android: Page Viewed (search)

# 🤔 Why

Datalake Event to Convert for Segment

When the user types a search term, the app currently makes a network call to fetch the results. When the results return from the API, call the Page Viewed event with the search term in the discover_search_term property with the correct total count in the discover_search_results_count

# 🛠 How

- Created method ``` trackSearchResultPageViewed ``` in ``` AnalyticEvents ```
```
/**
     * Sends data associated with the search results page viewed to segment.
     * @param discoveryParams: DiscoveryParams
     * @param count: Int
     * @param sort: DiscoveryParams.Sort
     */
    fun trackSearchResultPageViewed(discoveryParams: DiscoveryParams, count: Int, sort: DiscoveryParams.Sort) {
        val props = AnalyticEventsUtils.discoveryParamsProperties(discoveryParams).toMutableMap()
        props[CONTEXT_PAGE.contextName] = SEARCH.contextName
        props[DISCOVER_SORT.contextName] = sort.name.toLowerCase(Locale.ROOT)
        discoveryParams.term()?.let { props["discover_search_term"] = it }
        props["discover_search_results_count"] = count
        client.track(PAGE_VIEWED.eventName, props)
    }

```

- Called method from ``` SearchViewModel ``` on search results loaded from server
``` 
    params
          .compose(takePairWhen(this.discoverEnvelope))
          .compose(combineLatestPair(pageCount))
          .compose(bindToLifecycle())
          .filter(it -> ObjectUtils.isNotNull(it.first.first.term())
                  && StringExt.isPresent(it.first.first.term())
                  && it.first.first.sort() != defaultSort
                  && IntegerUtils.intValueOrZero(it.second) == 1)
          .distinct()
          .subscribe(it -> {
            this.lake.trackSearchResultPageViewed(defaultParams, it.first.second.stats().count(), defaultSort);
          });
```

- Added unit tests to segment event and properties 
``` SegmentTest ```
```
 @Test
    fun testSearchResultPageViewed_Properties() {
        val user = user()
        val client = client(user)
        client.eventNames.subscribe(this.segmentTrack)
        client.eventProperties.subscribe(this.propertiesTest)
        client.identifiedId.subscribe(this.segmentIdentify)
        val segment = AnalyticEvents(listOf(client))

        val params = DiscoveryParams
            .builder()
            .term("test")
            .sort(DiscoveryParams.Sort.POPULAR)
            .staffPicks(true)
            .build()

        segment.trackSearchResultPageViewed(params, 200, DiscoveryParams.Sort.POPULAR)

        assertSessionProperties(user)
        assertContextProperties()
        assertUserProperties(false)

        val expectedProperties = propertiesTest.value
        assertEquals("test", expectedProperties["discover_search_term"])
        assertEquals(200, expectedProperties["discover_search_results_count"])
        assertEquals(false, expectedProperties["discover_everything"])
        assertEquals(true, expectedProperties["discover_pwl"])
        assertEquals(false, expectedProperties["discover_recommended"])
        assertEquals("recommended_popular", expectedProperties["discover_ref_tag"])
        assertEquals(false, expectedProperties["discover_social"])
        assertEquals("popular", expectedProperties["discover_sort"])
        assertNull(expectedProperties["discover_subcategory_id"])
        assertNull(expectedProperties["discover_subcategory_name"])
        assertEquals(null, expectedProperties["discover_tag"])
        assertEquals(false, expectedProperties["discover_watched"])

        this.segmentTrack.assertValue(EventName.PAGE_VIEWED.eventName)
    }

```

``` SearchViewModelTest ```
```
this.lakeTest.assertValues("Search Button Clicked", EventName.CTA_CLICKED.getEventName(), "Search Page Viewed",  EventName.PAGE_VIEWED.getEventName());
```


# 👀 See

![Screenshot 2021-04-09 at 03 34 30](https://user-images.githubusercontent.com/63934292/114120837-33dd3580-98e5-11eb-9471-326c1ab30fbf.png)


| Before 🐛 | After 🦋 |
| --- | --- |
|  |  |

# 📋 QA

- Click on the search button from the global nav
- Type a search string into the search box, you should see the new event registered in segment console

<img width="368" alt="Screenshot 2021-04-09 at 03 36 47" src="https://user-images.githubusercontent.com/63934292/114120849-393a8000-98e5-11eb-9261-6f08740c1f82.png">


# Story 📖

https://kickstarter.atlassian.net/browse/EP-424
